[X86][NVPTX][LegalizeDAG] If i16 legal, legalize FABS/FNEG/FCOPYSIGN (f16) with Expand

v01dXYZ opened the pull request:
For the concerned targets (X86, NVPTX), `Expand` clears the sign bit after bitcasting to i16. This is cheaper than `Promote` that requires converting back and forth from f16 to f32 (with possibly expensive libcalls) in order to call FABS.f32

Fixes #104915.

Commit:
[X86][NVPTX][LegalizeDAG] If i16 legal, legalize FABS/FNEG/FCOPYSIGN (f16) with Expand

For the concerned targets, `Expand` clears/flips/copies the sign bit
after bitcasting to i16. This is cheaper than `Promote` that requires
converting back and forth from f16 to f32 (with possibly expensive
libcalls) in order to call FABS/FNEG/FCOPYSIGN (f32).

Reminder: By default, FCOPYSIGN action is Expand.

phoebewang reviewed (approved):
LGTM for X86 change, thanks!

Artem-B commented on llvm/test/CodeGen/NVPTX/f16x2-instructions.ll:
```diff
@@ -1183,17 +1183,13 @@ define <2 x half> @test_fma(<2 x half> %a, <2 x half> %b, <2 x half> %c) #0 {
-; CHECK:      st.param.b32    [func_retval0+0], [[R]];
-; CHECK:      ret;
+; CHECK:    ld.param.b32    [[A:%r[0-9]+]], [test_fabs_param_0];
+; CHECK:    mov.b32         {[[A0:%rs[0-9]+]], [[A1:%rs[0-9]+]]}, [[A]]
+; CHECK:    and.b16 [[A2:%rs[0-9]+]], [[A1]], 32767;
+; CHECK:    and.b16 [[A3:%rs[0-9]+]], [[A0]], 32767;
+; CHECK:    mov.b32 [[B:%r[0-9]+]], {[[A3]], [[A2]]};
```
Technically this could be improved further by performing `and` on the i32 directly. I suspect ptxas may do that for us.

Maybe add a TODO note.

gonzalobg replied:
I think a TODO is fine, but that we should not rely on `ptxas` doing it for us.

v01dXYZ replied:
LegalizeType would type-expand `FABS` (`<2 x f16>`).

The DAG just before instruction selection:

```
Optimized legalized selection DAG: %bb.0 'test_fabs:'
SelectionDAG has 21 nodes:
  t0: ch,glue = EntryToken
  t14: v2f16,ch = load<(dereferenceable invariant load (s32) from `ptr addrspace(101) null`, addrspace 101)> t0, TargetExternalSymbol:i64'test_fabs_param_0', undef:i64
      t8: ch = CopyToReg t0, Register:v2f16 %0, t14
              t16: f16 = extract_vector_elt t14, Constant:i64<0>
            t22: i16 = bitcast t16
          t24: i16 = and t22, Constant:i16<32767>
        t25: f16 = bitcast t24
              t19: f16 = extract_vector_elt t14, Constant:i64<1>
            t26: i16 = bitcast t19
          t27: i16 = and t26, Constant:i16<32767>
        t28: f16 = bitcast t27
      t21: v2f16 = BUILD_VECTOR t25, t28
    t11: ch = NVPTXISD::StoreRetval<(store (s32), align 1)> t8, Constant:i32<0>, t21
  t12: ch = NVPTXISD::RET_GLUE t11
``` 

Maybe detect a composition of bitwise operations on each element and merge it into a composition on the bitcasted vector itself (starting from `BUILD_VECTOR`).

BTW that's also the case with `x86`:

``` llvm
define <2 x half> @test_fabs(<2 x half> %a) #0 {
  %r = call <2 x half> @llvm.fabs.f16(<2 x half> %a)
  ret <2 x half> %r
}
``` 

``` asm
        .text
        .file   "fabs_nvptx.ll"
        .globl  test_fabs                       # -- Begin function test_fabs
        .p2align        4, 0x90
        .type   test_fabs,@function
test_fabs:                              # @test_fabs
        .cfi_startproc
# %bb.0:
        pextrw  $0, %xmm0, %eax
        psrld   $16, %xmm0
        pextrw  $0, %xmm0, %ecx
        andl    $32767, %ecx                    # imm = 0x7FFF
        pinsrw  $0, %ecx, %xmm1
        andl    $32767, %eax                    # imm = 0x7FFF
        pinsrw  $0, %eax, %xmm0
        punpcklwd       %xmm1, %xmm0            # xmm0 = xmm0[0],xmm1[0],xmm0[1],xmm1[1],xmm0[2],xmm1[2],xmm0[3],xmm1[3]
        retq
.Lfunc_end0:
        .size   test_fabs, .Lfunc_end0-test_fabs
        .cfi_endproc
                                        # -- End function
        .section        ".note.GNU-stack","",@progbits
```

But `AArch64`  seems not to do so (there is a combiner that combine the build_vector into a `concat_vector ..., undef`).

Artem-B replied:
Let's not get sidetracked. This is something for a separate patch or github issue, and would be best discussed there.

A comment highlighting a future optimization opportunity is all that we need in this patch.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -849,7 +849,8 @@ NVPTXTargetLowering::NVPTXTargetLowering(const NVPTXTargetMachine &TM,
   for (const auto &Op : {ISD::FABS}) {
-    setOperationAction(Op, MVT::f16, Promote);
+    // Expand instead of Promote to clear sign bit by bitcasting to i16
+    setOperationAction(Op, MVT::f16, Expand);
```
Would operations other than FABS benefit from that, too?  E.g. `fneg` or `fcopysign`?

v01dXYZ replied:
that's a valid point.

For `NVPTX`, we have for `f16` (so no change required)

* `FNEG` Legal / Expand
* `FCOPYSIGN` Expand

For `X86-64`, we have for `f16`

* `FNEG` Promote
* `FCOPYSIGN` Expand

Artem-B reviewed (approved):
LGTM.

v01dXYZ commented:
@phoebewang I added also `FNEG`/`FCOPYSIGN`. I rewrote a little bit the code to avoid having special cases within `setF16Action` (not more efficient but simpler).

Artem-B commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -672,6 +672,11 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
+    // Expand instead of Promote to clear/flip/copy sign bit by bitcasting to
+    // i16.
+    setOperationAction(ISD::FABS, MVT::f16, Expand);
+    setOperationAction(ISD::FNEG, MVT::f16, Expand);
```
X86 could use test cases for fneg/fcopysign.

v01dXYZ replied:
Thanks !

phoebewang commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -601,7 +601,7 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
     setOperationAction(ISD::FABS, VT, Action);
     setOperationAction(ISD::FNEG, VT, Action);
-    setOperationAction(ISD::FCOPYSIGN, VT, Expand);
+    setOperationAction(ISD::FCOPYSIGN, VT, Action);
```
What's the reason to change this? And why don't set above Expand here?

v01dXYZ replied:
Because it would create even more special cases that don't reflect the function name.

The function is called elsewhere with `Expand` unless it's `f16`.

That's just for consistency, I could do it here if it's better.

phoebewang replied:
This function is called not only by scalar `f16` but also both vector FP16 and BF16, though we don't have tests to cover them.

v01dXYZ replied:
When it is for a vector FP/BF, it is always called with the Expand Operation.

phoebewang replied:
Thanks, then it's fine for me :)

phoebewang replied:
Wait.. should we remove it from `setF16Action` instead?

phoebewang replied:
> Wait.. should we remove it from `setF16Action` instead?

Sorry, I misunderstood. No problem then.

v01dXYZ replied:
1. we can isolate those Opcodes into another function.
2. keep them and overwrite the actions just after calling `setF16Action` if that's not the right ones (which is the path I've taken, and that's what's in the current code with `MVT::v16f16` for example).

phoebewang replied:
I was thinking you meant `FCOPYSIGN` is set Expand by default, then I got you mean it's all set Expand in X86ISelLowering.cpp.
Just have a double check, it does set Expand by default, see
https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/TargetLoweringBase.cpp#L775
So we can simply remove it from `setF16Action`.

v01dXYZ replied:
That's right. The less code the better.

RKSimon reviewed (changes requested):
please can you rebase? I've already added additional test coverage to fp16-libcalls.ll, including copysign/fabs/fneg etc.

RKSimon reviewed (approved):
LGTM - cheers

arsenm reviewed (commented):
This is still papering over a bug - the promote action is just busted and needs to be replaced or removed

v01dXYZ commented:
@arsenm If all targets with f16 have a valid i16, it would mean that the natural way to lower those ops is with a bitcast (unless there are native instructions for that of course). In this case, removing Promote is sound.

Concerning the case of an arch with f16 and without i16, we should instead use the smallest fitting integer type instead of promotion.

So yes, promote seems to be overkill for operations that could be implemented as bitwise ops after move back and forth to a GPR.

arsenm commented:
> * For targets with f16 have a valid i16, the natural way to lower those ops is with a bitcast (unless there are native instructions for that of course). In this case, removing Promote is sound.

The problem isn't that it's overkill, it's that it is *wrong*. You cannot have the canonicalization effects of the fpextend.

> * Concerning the case of an arch with f16 and without i16, we should instead use the smallest fitting integer type rather than promotion.

This is more of a ridiculous DAG design issue; you should be able to create the bitcast and have later legalization steps deal with it. Failing that, you could do a replacement bitcast by doing store to stack and reload

v01dXYZ commented:
Indeed, [LangRef](https://llvm.org/docs/LangRef.html#llvm-fabs-intrinsic) clearly mentions the fact that only the bit sign is changed with fabs/fneg/fcopysign but say nothing about that with fpext/fptrunc.

RKSimon reviewed (changes requested):
This appears to have rotted - please can you merge/rebase against trunk latest?
